ASoC: SOF: Intel/ipc4:  Follow the hardware proramming sequence for library loading

ujfalusi opened the pull request:
The recommended programming sequence for the HD-DMA is to first configure
the buffer and set the GEN bit on the DSP side and enable the host DMA
with the RUN bit.
If this sequence is not followed the DMA might not start copying data or
it can lead to unpredictable behavior.

lyakh commented on sound/soc/sof/intel/hda-ctrl.c:
```diff
@@ -173,8 +173,8 @@ int hda_dsp_ctrl_clock_power_gating(struct snd_sof_dev *sdev, bool enable)
-	val = enable ? 0 : PCI_PGCTL_ADSPPGD;
-	snd_sof_pci_update_bits(sdev, PCI_PGCTL, PCI_PGCTL_ADSPPGD, val);
+	val = enable ? 0 : (PCI_PGCTL_ADSPPGD | BIT(1));
+	snd_sof_pci_update_bits(sdev, PCI_PGCTL, PCI_PGCTL_ADSPPGD | BIT(1), val);
```
should we add a macro for `BIT(1)`?

ranj063 replied:
is this change really necessary?

ujfalusi replied:
It is not needed, leftover from the million rounds of trying to figure out what is going on..

lyakh commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -551,22 +589,25 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
+	if (loader_data.dma_started) {
+		/*
+		 * The DMA was started in hda_dsp_ipc4_load_library_dma_trigger()
+		 * callback during the message sending
```
maybe add a phrase here - "have to stop it regardless whether we received a reply from the firmware or not"

ujfalusi replied:
The DMA has to be stopped if it was started, the comment helps the reader who does not see when it is started in this function and might get confused.

lyakh reviewed (approved):
minor comments only

ranj063 commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -668,7 +668,7 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
 					 sd_offset +
 					 SOF_HDA_ADSP_REG_SD_FIFOSIZE);
-		hstream->fifo_size &= 0xffff;
+		hstream->fifo_size &= SOF_HDA_SD_FIFOSIZE_FIFOS_MASK;
```
can we squash this with the previous commit?

ujfalusi replied:
Yes, we can

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -551,22 +589,25 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
+		}
 	}
 
-cleanup:
```
why do we need to cleanup is loader_data.dma_started was never true?

ujfalusi replied:
We always need to do the cleanup after hda_cl_stream_prepare(), the cleanup is needed regardless if the start failed (dma is not running).

ujfalusi commented:
Changes since v1:
- Squash the SOF_HDA_SD_FIFOSIZE_FIFOS_MASK introduction and use of it in hda-stream.c
- Update the comment before stopping the DMA channel
- Drop the leftover experiment in hda-ctrl.c (power/clock gating change)

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -28,6 +30,11 @@
+ *
+ * @callback_after_send: Optional callback function to be called after the
+ *			 message has been sent and before starting to wait for
+ *			 the response from the firmware
```
the description and commit message don't even begin to describe why this is needed?

what could possibly be done between sending an IPC and waiting for the result?

And how is that a 'split' handling? Once the message has been sent, what can we really control?

ujfalusi replied:
This is a generic callback with a clear definition on how it is going to be called and when exactly.
At this point I don't think it is right to narrow it down to one specific use which is at the end the reason why it is introduced.
At this point of time I don't know if this callback is going to be used for other purposes, we did not even knew that this is need at all before SOF firmware implemented the library loading...

plbossart commented on sound/soc/sof/ipc4.c:
```diff
@@ -352,6 +352,9 @@ static int ipc4_tx_msg_unlocked(struct snd_sof_ipc *ipc,
 	}
 
+	if (ipc4_msg->callback_after_send)
+		ipc4_msg->callback_after_send(sdev, ipc4_msg->callback_data);
```
the commit message should describe what happens if the firmware replies immediately before the callback_after_send() is invoked.

Waiting for the firmware is required, but what if the firmware replies very quickly?

Also why can't we invoke the callback AFTER the IPC response is handled?

ujfalusi replied:
The split handling or specific flow is:
Kernel sends _one_ message
Firmware does something
Kernel needs to do something to follow a programming sequence for the operation to be executed
Firmware sends response as the operation will complete.

We cannot invoke the callback AFTER the IPC response as the response will not come unless we execute the callback after sending the message.

In the specific case we are aiming at this point is the LOAD_LIBRARY.
The programming sequence for it is:
1. Firmware buffer setup + GEN bit set
2. Kernel sets RUN bit for the channel

We have a single message: LOAD_LIBRARY
This message will make the DSP to setup it's buffer and set the GEN bit and _wait_ for the data to be transferred.
If we set the RUN bit in kernel **before** the message: programming sequence violation, unsupported flow
if we would set the RUN bit **after** the message: the IPC times out (no reply as the DMA is not running, data will not be transferred) setting the RUN bit after the firmware failed the message does not make sense and invalid.

plbossart replied:
"We cannot invoke the callback AFTER the IPC response as the response will not come unless we execute the callback after sending the message."

That's the part you have not described at all in any of your commit messages or comments.
This is a special case of an IPC needing an extra configuration before firmware can reply.

"callback_after_send" does not even begin to describe what you are trying to do. "after_send" could be understood as "after pushing the IPC button, before the reply" or "after completing an IPC, reply included".

ujfalusi replied:
OK, I will try to reword the commit messages and the comment for the `callback_after_send`, for which I don't have a better name in my mind atm.

plbossart replied:
"callback_before_ipc_reply" ? It's literally what the design intent is.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -516,10 +516,48 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
+
+	/*
+	 * Wait for firmware to set the GEN bit, the SDxFIFOS register FIFO size
+	 * will change to a non zero value.
```
that's exactly the sort of comment that's missing from the initial patches. Here there is an explicit wait on a host-visible register before handling the IPC response.

But that doesn't tell me why we couldn't do this *AFTER* the IPC response is received, which would make more sense to me.

The commit message is equally confusing: 

"
In order to comply with the hardware programming flow, use the optional
callback_after_send to start the DMA after sending the LOAD_LIBRARY
message.
"

-> so why would we involved the callback before the LOAD_LIBRARY IPC response is received.

I am really not getting the concept of doing something without waiting for the IPC response. What if the IPC times out, then we would have done something that will not work anyways.

ujfalusi replied:
The information is here because it is relevant in this patch.

We cannot do AFTER the IPC response as there is no IPC response if the DMA is not running.

The programming flow for the DMA channel must be:
1. firmware sets up it's buffer and sets the GEN bit
2. kernel sets the RUN bit.

With this implementation we will:
A. kernel sends IPC message
B. firmware configures the buffer and sets the GEN bit (**STEP 1**)
C. firmware enters waiting for the library data via DMA
D. the kernel in the callback waits for the FIFOS to be != 0 (the GEN bit set will trigger this in hardware) and sets the RUN bit (**STEP 2**)
E. firmware receives the whole library data and sends response (clearing also GEN bit)
F. kernel receives the response and clears the RUN bit

ujfalusi reviewed (commented):
@plbossart,

The original sequence which was verified (and re-verified again and again) with the reference firmware is:
1. Load basefw
2. boot basefw and wait for FW_READY
3. start host DMA for a library to be loaded
4. send LOAD_LIBRARY to firmware
5. firmware receives the library via DMA and sends the response
6. library is loaded and working

With SOF implementing the support:
1. Load basefw
2. boot basefw and wait for FW_READY
3. start host DMA for a library to be loaded
4. send LOAD_LIBRARY to firmware
5. the message is timing out
6. firmware ends up in endless loop

Looking at the flow, one thing is clear: the programming sequence is against the supported flow, so in any ways we need to correct it.

As for WHY the reference firmware works and SOF does not? I really don't have answers and trying to get more information on it internally so we can know exactly what is going on and if we should fix this in the firmware somehow.

ujfalusi commented:
Closing as it was decided to introduce a new message for the split use as preparation stage.
I have the working set but we will need firmware side support (and name for the new message).
